cli/src/main.rs

```diff
@@ -1,19 +1,561 @@
+    });
+
+    info!("Started development server");
+    debug!("HTTP development server listening on {}", http_addr);
+    debug!("RPC server listening on {}", rpc_addr);
+
+    if *open_browser {
+        let server_state = poll_server_state(domain, rpc_port, &true).await;
+        if server_state.is_ok_and(|s| s == ServerState::Started) {
+            match open::that(&http_addr) {
+                Ok(()) => debug!("Opened '{}' in your default browser.", http_addr),
+                Err(err) => error!("An error occurred when opening '{}': {}", http_addr, err),
+            }
+        }
+    }
+
+    info!("You can reach the development server on {}", http_addr);
+
+    return tokio::select! {
+        // Wait for the handlers to exit. Currently this will never happen
+        wait = wait => trace!("wait {:?}", wait),
+        task_handle = task_handle => {
+            match task_handle {
+                Ok(_) => std::process::exit(-1),
+                Err(e) => error!("task_handle error {:?}", e)
+            }
+        }
+    };
+}
+
+fn start_development_server(
+    http_port: &u16,
+    rpc_port: &u16,
+) -> (DevelopmentServerClient, impl Future<Output = ()>) {
+    let cargo = env::var("CARGO").unwrap_or_else(|_| "cargo".to_string());
+    let log_level = env::var("RUST_LOG").expect("env::var RUST_LOG not set");
+    let mut process = Command::new(cargo)
+        .env("RUST_LOG", log_level)
+        .current_dir(project_root())
+        .args(&[
+            "run",
+            "--package=development_server",
+            "--",
+            format!("--http-port={}", http_port).as_str(),
+            format!("--rpc-port={}", rpc_port).as_str(),
+        ])
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .kill_on_drop(true)
+        .spawn()
+        .expect("Unable to spawn process");
+
+    let proc_id = process.id().expect("Unable to fetch process id");
+    trace!("proc id: {}", proc_id);
+
+    let stdout = process.stdout.take().unwrap();
+    let stdout_reader = BufReader::new(stdout);
+
+    let stderr = process.stderr.take().unwrap();
+    let stderr_reader = BufReader::new(stderr);
+
+    // This future waits for handlers to exit, we don't want to await it here.
+    // Return it instead so the caller can await it.
+    let wait = setup_listeners(stdout_reader, stderr_reader);
+
+    // Send initialize request
+    debug!("initialize request");
+    let client = DevelopmentServerClient { process };
+
+    (client, wait)
+}
+
+async fn setup_listeners(
+    stdout_reader: BufReader<ChildStdout>,
+    stderr_reader: BufReader<ChildStderr>,
+) {
+    let handle_stdout = tokio::spawn(async move {
+        let mut reader = stdout_reader.lines();
+        loop {
+            match reader.next_line().await {
+                Ok(Some(string)) => trace!("handle_stdout: {}", string),
+                Ok(None) => trace!("handle_stdout: None"),
+                Err(e) => trace!("handle_stdout: {:?}", e),
```

Comment:
I think we might benefit by using a `error!` for the error cases here. Same with the following loop

Reply:
I don't mind doing this. But this results in the following output when stopping a running server from another terminal session using the default log level:

```terminal
mba~/code/dl/mycelia % cargo run start
   Compiling cli v0.1.0 (/Users/dani/code/dl/mycelia/cli)
    Finished dev [unoptimized + debuginfo] target(s) in 2.41s
     Running `target/debug/cli start`
[2023-09-25T15:48:48Z INFO  cli] Starting development server
[2023-09-25T15:48:48Z INFO  cli] Started development server
[2023-09-25T15:48:50Z WARN  cli] Development server already listening
[2023-09-25T15:48:50Z INFO  cli] You can reach the development server on http://127.0.0.1:3001
[2023-09-25T15:48:50Z INFO  cli] handle_stderr:     Finished dev [unoptimized + debuginfo] target(s) in 0.33s
[2023-09-25T15:48:50Z INFO  cli] handle_stderr:      Running `target/debug/development_server --http-port=3001 --rpc-port=50051`
[2023-09-25T15:48:50Z INFO  cli] handle_stderr: [2023-09-25T15:48:49Z INFO  development_server] starting up
[2023-09-25T15:48:50Z INFO  cli] handle_stderr: [2023-09-25T15:48:49Z INFO  development_server::rpc] preparing to start rpc server
[2023-09-25T15:48:50Z INFO  cli] handle_stderr: [2023-09-25T15:48:50Z INFO  development_server::http_function_component] Starting HTTP Server
[2023-09-25T15:48:50Z INFO  cli] handle_stderr: [2023-09-25T15:48:50Z INFO  development_server::rpc] Received Echo. message: "cli::server_state()"
[2023-09-25T15:49:13Z INFO  cli] handle_stderr: [2023-09-25T15:49:13Z WARN  development_server::http_function_component] command loop handle returned
[2023-09-25T15:49:13Z INFO  cli] handle_stderr: [2023-09-25T15:49:13Z WARN  development_server] http server task completed
[2023-09-25T15:49:13Z INFO  cli] handle_stdout: None
[2023-09-25T15:49:13Z INFO  cli] handle_stdout: None
[2023-09-25T15:49:13Z INFO  cli] handle_stdout: None
[2023-09-25T15:49:13Z INFO  cli] handle_stderr: None
[2023-09-25T15:49:13Z INFO  cli] handle_stdout: None
[2023-09-25T15:49:13Z INFO  cli] handle_stdout: None
[2023-09-25T15:49:13Z INFO  cli] handle_stderr: None
[2023-09-25T15:49:13Z INFO  cli] handle_stdout: None
[2023-09-25T15:49:13Z INFO  cli] handle_stderr: None
[2023-09-25T15:49:13Z INFO  cli] handle_stdout: None
[2023-09-25T15:49:13Z INFO  cli] handle_stdout: None
[2023-09-25T15:49:13Z INFO  cli] handle_stderr: None
[2023-09-25T15:49:13Z INFO  cli] handle_stdout: None
[2023-09-25T15:49:13Z INFO  cli] handle_stdout: None
[2023-09-25T15:49:13Z INFO  cli] handle_stderr: None
[2023-09-25T15:49:13Z INFO  cli] handle_stdout: None
[2023-09-25T15:49:13Z INFO  cli] handle_stdout: None
[%                                                                                                                                                       
```

as opposed to:

```terminal
mba~/code/dl/mycelia % cargo run start
    Finished dev [unoptimized + debuginfo] target(s) in 0.22s
     Running `target/debug/cli start`
[2023-09-25T15:50:41Z INFO  cli] Starting development server
[2023-09-25T15:50:41Z INFO  cli] Started development server
[2023-09-25T15:50:43Z WARN  cli] Development server already listening
[2023-09-25T15:50:43Z INFO  cli] You can reach the development server on http://127.0.0.1:3001
[2023-09-25T15:50:43Z INFO  cli] handle_stderr:     Finished dev [unoptimized + debuginfo] target(s) in 0.32s
[2023-09-25T15:50:43Z INFO  cli] handle_stderr:      Running `target/debug/development_server --http-port=3001 --rpc-port=50051`
[2023-09-25T15:50:43Z INFO  cli] handle_stderr: [2023-09-25T15:50:42Z INFO  development_server] starting up
[2023-09-25T15:50:43Z INFO  cli] handle_stderr: [2023-09-25T15:50:42Z INFO  development_server::rpc] preparing to start rpc server
[2023-09-25T15:50:43Z INFO  cli] handle_stderr: [2023-09-25T15:50:43Z INFO  development_server::http_function_component] Starting HTTP Server
[2023-09-25T15:50:43Z INFO  cli] handle_stderr: [2023-09-25T15:50:43Z INFO  development_server::rpc] Received Echo. message: "cli::server_state()"
[2023-09-25T15:50:46Z INFO  cli] handle_stderr: [2023-09-25T15:50:46Z WARN  development_server::http_function_component] command loop handle returned
[2023-09-25T15:50:46Z INFO  cli] handle_stderr: [2023-09-25T15:50:46Z WARN  development_server] http server task completed
[2023-09-25T15:50:46Z ERROR cli] Process exited. Cannot continue. Error:
    
    ExitStatus(
        unix_wait_status(
            0,
        ),
    )
```

I'll change it to info if that still has your preference, but I found this a bit messy and not really adding any valuable info. Hence I changed it to trace

Reply:
oh goodness. Please keep it as is lol. Didn't realize it would produce so much spam!